File: Source/DafnyStandardLibraries/binaries/Python/translated-py/Std_JSON_ZeroCopy_Deserializer_ObjectParams.py
import sys
from typing import Callable, Any, TypeVar, NamedTuple
from math import floor
from itertools import count

import module_
import _dafny
import System_
import Std_Wrappers
import Std_BoundedInts
import Std_Base64
import Std_Relations
import Std_Math
import Std_Collections_Seq
import Std_Collections_Array
import Std_Collections_Imap
import Std_Functions
import Std_Collections_Iset
import Std_Collections_Map
import Std_Collections_Set
import Std_Collections
import Std_DynamicArray
import Std_FileIO
import Std_Arithmetic_GeneralInternals
import Std_Arithmetic_MulInternalsNonlinear
import Std_Arithmetic_MulInternals
import Std_Arithmetic_Mul
import Std_Arithmetic_ModInternalsNonlinear
import Std_Arithmetic_DivInternalsNonlinear
import Std_Arithmetic_ModInternals
import Std_Arithmetic_DivInternals
import Std_Arithmetic_DivMod
import Std_Arithmetic_Power
import Std_Arithmetic_Logarithm
import Std_Arithmetic_Power2
import Std_Arithmetic
import Std_Strings_HexConversion
import Std_Strings_DecimalConversion
import Std_Strings_CharStrEscaping
import Std_Strings
import Std_Unicode_Base
import Std_Unicode_Utf8EncodingForm
import Std_Unicode_Utf16EncodingForm
import Std_Unicode_UnicodeStringsWithUnicodeChar
import Std_Unicode_Utf8EncodingScheme
import Std_Unicode
import Std_PythonConcurrent
import Std_PythonFileIOInternalExterns
import Std_JSON_Values
import Std_JSON_Errors
import Std_JSON_Spec
import Std_JSON_Utils_Views_Core
import Std_JSON_Utils_Views_Writers
import Std_JSON_Utils_Views
import Std_JSON_Utils_Lexers_Core
import Std_JSON_Utils_Lexers_Strings
import Std_JSON_Utils_Lexers
import Std_JSON_Utils_Cursors
import Std_JSON_Utils_Parsers
import Std_JSON_Utils
import Std_JSON_Grammar
import Std_JSON_ByteStrConversion
import Std_JSON_Serializer
import Std_JSON_Deserializer_Uint16StrConversion
import Std_JSON_Deserializer
import Std_JSON_ConcreteSyntax_Spec
import Std_JSON_ConcreteSyntax_SpecProperties
import Std_JSON_ConcreteSyntax
import Std_JSON_ZeroCopy_Serializer
import Std_JSON_ZeroCopy_Deserializer_Core
import Std_JSON_ZeroCopy_Deserializer_Strings
import Std_JSON_ZeroCopy_Deserializer_Numbers

# Module: Std_JSON_ZeroCopy_Deserializer_ObjectParams

class default__:
    def  __init__(self):
        pass

    @staticmethod
    def Colon(cs):
        d_647_valueOrError0_ = (cs).AssertChar(_dafny.CodePoint(':'))
        if (d_647_valueOrError0_).IsFailure():
            return (d_647_valueOrError0_).PropagateFailure()
        elif True:
            d_648_cs_ = (d_647_valueOrError0_).Extract()
            return Std_Wrappers.Result_Success((d_648_cs_).Split())

    @staticmethod
    def KeyValueFromParts(k, colon, v):
        d_649_sp_ = Std_JSON_Utils_Cursors.Split_SP(Std_JSON_Grammar.jKeyValue_KeyValue((k).t, (colon).t, (v).t), (v).cs)
        return d_649_sp_

    @staticmethod
    def ElementSpec(t):
        return Std_JSON_ConcreteSyntax_Spec.default__.KeyValue(t)

    @staticmethod
    def Element(cs, json):
        d_650_valueOrError0_ = Std_JSON_ZeroCopy_Deserializer_Strings.default__.String(cs)
        if (d_650_valueOrError0_).IsFailure():
            return (d_650_valueOrError0_).PropagateFailure()
        elif True:
            d_651_k_ = (d_650_valueOrError0_).Extract()
            d_652_p_ = default__.Colon
            d_653_valueOrError1_ = Std_JSON_ZeroCopy_Deserializer_Core.default__.Structural((d_651_k_).cs, d_652_p_)
            if (d_653_valueOrError1_).IsFailure():
                return (d_653_valueOrError1_).PropagateFailure()
            elif True:
                d_654_colon_ = (d_653_valueOrError1_).Extract()
                d_655_valueOrError2_ = (json)((d_654_colon_).cs)
                if (d_655_valueOrError2_).IsFailure():
                    return (d_655_valueOrError2_).PropagateFailure()
                elif True:
                    d_656_v_ = (d_655_valueOrError2_).Extract()
                    d_657_kv_ = default__.KeyValueFromParts(d_651_k_, d_654_colon_, d_656_v_)
                    return Std_Wrappers.Result_Success(d_657_kv_)

    @_dafny.classproperty
    def OPEN(instance):
        return ord(_dafny.CodePoint('{'))
    @_dafny.classproperty
    def CLOSE(instance):
        return ord(_dafny.CodePoint('}'))

File: Source/DafnyRuntime/DafnyRuntimeGo/output/Std_JSON_ZeroCopy_Deserializer_ObjectParams/Std_JSON_ZeroCopy_Deserializer_ObjectParams.go
// Package Std_JSON_ZeroCopy_Deserializer_ObjectParams
// Dafny module Std_JSON_ZeroCopy_Deserializer_ObjectParams compiled into Go

package Std_JSON_ZeroCopy_Deserializer_ObjectParams

import (
  _dafny "dafny"
  os "os"
  _System "System_"
  Std_Wrappers "Std_Wrappers"
  Std_Concurrent "Std_Concurrent"
  Std_FileIOInternalExterns "Std_FileIOInternalExterns"
  Std_BoundedInts "Std_BoundedInts"
  Std_Base64 "Std_Base64"
  Std_Relations "Std_Relations"
  Std_Math "Std_Math"
  Std_Collections_Seq "Std_Collections_Seq"
  Std_Collections_Array "Std_Collections_Array"
  Std_Collections_Imap "Std_Collections_Imap"
  Std_Functions "Std_Functions"
  Std_Collections_Iset "Std_Collections_Iset"
  Std_Collections_Map "Std_Collections_Map"
  Std_Collections_Set "Std_Collections_Set"
  Std_Collections "Std_Collections"
  Std_DynamicArray "Std_DynamicArray"
  Std_FileIO "Std_FileIO"
  Std_Arithmetic_GeneralInternals "Std_Arithmetic_GeneralInternals"
  Std_Arithmetic_MulInternalsNonlinear "Std_Arithmetic_MulInternalsNonlinear"
  Std_Arithmetic_MulInternals "Std_Arithmetic_MulInternals"
  Std_Arithmetic_Mul "Std_Arithmetic_Mul"
  Std_Arithmetic_ModInternalsNonlinear "Std_Arithmetic_ModInternalsNonlinear"
  Std_Arithmetic_DivInternalsNonlinear "Std_Arithmetic_DivInternalsNonlinear"
  Std_Arithmetic_ModInternals "Std_Arithmetic_ModInternals"
  Std_Arithmetic_DivInternals "Std_Arithmetic_DivInternals"
  Std_Arithmetic_DivMod "Std_Arithmetic_DivMod"
  Std_Arithmetic_Power "Std_Arithmetic_Power"
  Std_Arithmetic_Logarithm "Std_Arithmetic_Logarithm"
  Std_Arithmetic_Power2 "Std_Arithmetic_Power2"
  Std_Arithmetic "Std_Arithmetic"
  Std_Strings_HexConversion "Std_Strings_HexConversion"
  Std_Strings_DecimalConversion "Std_Strings_DecimalConversion"
  Std_Strings_CharStrEscaping "Std_Strings_CharStrEscaping"
  Std_Strings "Std_Strings"
  Std_Unicode_Base "Std_Unicode_Base"
  Std_Unicode_Utf8EncodingForm "Std_Unicode_Utf8EncodingForm"
  Std_Unicode_Utf16EncodingForm "Std_Unicode_Utf16EncodingForm"
  Std_Unicode_UnicodeStringsWithUnicodeChar "Std_Unicode_UnicodeStringsWithUnicodeChar"
  Std_Unicode_Utf8EncodingScheme "Std_Unicode_Utf8EncodingScheme"
  Std_Unicode "Std_Unicode"
  Std_JSON_Values "Std_JSON_Values"
  Std_JSON_Errors "Std_JSON_Errors"
  Std_JSON_Spec "Std_JSON_Spec"
  Std_JSON_Utils_Views_Core "Std_JSON_Utils_Views_Core"
  Std_JSON_Utils_Views_Writers "Std_JSON_Utils_Views_Writers"
  Std_JSON_Utils_Views "Std_JSON_Utils_Views"
  Std_JSON_Utils_Lexers_Core "Std_JSON_Utils_Lexers_Core"
  Std_JSON_Utils_Lexers_Strings "Std_JSON_Utils_Lexers_Strings"
  Std_JSON_Utils_Lexers "Std_JSON_Utils_Lexers"
  Std_JSON_Utils_Cursors "Std_JSON_Utils_Cursors"
  Std_JSON_Utils_Parsers "Std_JSON_Utils_Parsers"
  Std_JSON_Utils "Std_JSON_Utils"
  Std_JSON_Grammar "Std_JSON_Grammar"
  Std_JSON_ByteStrConversion "Std_JSON_ByteStrConversion"
  Std_JSON_Serializer "Std_JSON_Serializer"
  Std_JSON_Deserializer_Uint16StrConversion "Std_JSON_Deserializer_Uint16StrConversion"
  Std_JSON_Deserializer "Std_JSON_Deserializer"
  Std_JSON_ConcreteSyntax_Spec "Std_JSON_ConcreteSyntax_Spec"
  Std_JSON_ConcreteSyntax_SpecProperties "Std_JSON_ConcreteSyntax_SpecProperties"
  Std_JSON_ConcreteSyntax "Std_JSON_ConcreteSyntax"
  Std_JSON_ZeroCopy_Serializer "Std_JSON_ZeroCopy_Serializer"
  Std_JSON_ZeroCopy_Deserializer_Core "Std_JSON_ZeroCopy_Deserializer_Core"
  Std_JSON_ZeroCopy_Deserializer_Strings "Std_JSON_ZeroCopy_Deserializer_Strings"
  Std_JSON_ZeroCopy_Deserializer_Numbers "Std_JSON_ZeroCopy_Deserializer_Numbers"
)
var _ _dafny.Dummy__
var _ = os.Args
var _ _System.Dummy__
var _ Std_Wrappers.Dummy__
var _ Std_Concurrent.Dummy__
var _ Std_FileIOInternalExterns.Dummy__
var _ Std_BoundedInts.Dummy__
var _ Std_Base64.Dummy__
var _ Std_Relations.Dummy__
var _ Std_Math.Dummy__
var _ Std_Collections_Seq.Dummy__
var _ Std_Collections_Array.Dummy__
var _ Std_Collections_Imap.Dummy__
var _ Std_Functions.Dummy__
var _ Std_Collections_Iset.Dummy__
var _ Std_Collections_Map.Dummy__
var _ Std_Collections_Set.Dummy__
var _ Std_Collections.Dummy__
var _ Std_DynamicArray.Dummy__
var _ Std_FileIO.Dummy__
var _ Std_Arithmetic_GeneralInternals.Dummy__
var _ Std_Arithmetic_MulInternalsNonlinear.Dummy__
var _ Std_Arithmetic_MulInternals.Dummy__
var _ Std_Arithmetic_Mul.Dummy__
var _ Std_Arithmetic_ModInternalsNonlinear.Dummy__
var _ Std_Arithmetic_DivInternalsNonlinear.Dummy__
var _ Std_Arithmetic_ModInternals.Dummy__
var _ Std_Arithmetic_DivInternals.Dummy__
var _ Std_Arithmetic_DivMod.Dummy__
var _ Std_Arithmetic_Power.Dummy__
var _ Std_Arithmetic_Logarithm.Dummy__
var _ Std_Arithmetic_Power2.Dummy__
var _ Std_Arithmetic.Dummy__
var _ Std_Strings_HexConversion.Dummy__
var _ Std_Strings_DecimalConversion.Dummy__
var _ Std_Strings_CharStrEscaping.Dummy__
var _ Std_Strings.Dummy__
var _ Std_Unicode_Base.Dummy__
var _ Std_Unicode_Utf8EncodingForm.Dummy__
var _ Std_Unicode_Utf16EncodingForm.Dummy__
var _ Std_Unicode_UnicodeStringsWithUnicodeChar.Dummy__
var _ Std_Unicode_Utf8EncodingScheme.Dummy__
var _ Std_Unicode.Dummy__
var _ Std_JSON_Values.Dummy__
var _ Std_JSON_Errors.Dummy__
var _ Std_JSON_Spec.Dummy__
var _ Std_JSON_Utils_Views_Core.Dummy__
var _ Std_JSON_Utils_Views_Writers.Dummy__
var _ Std_JSON_Utils_Views.Dummy__
var _ Std_JSON_Utils_Lexers_Core.Dummy__
var _ Std_JSON_Utils_Lexers_Strings.Dummy__
var _ Std_JSON_Utils_Lexers.Dummy__
var _ Std_JSON_Utils_Cursors.Dummy__
var _ Std_JSON_Utils_Parsers.Dummy__
var _ Std_JSON_Utils.Dummy__
var _ Std_JSON_Grammar.Dummy__
var _ Std_JSON_ByteStrConversion.Dummy__
var _ Std_JSON_Serializer.Dummy__
var _ Std_JSON_Deserializer_Uint16StrConversion.Dummy__
var _ Std_JSON_Deserializer.Dummy__
var _ Std_JSON_ConcreteSyntax_Spec.Dummy__
var _ Std_JSON_ConcreteSyntax_SpecProperties.Dummy__
var _ Std_JSON_ConcreteSyntax.Dummy__
var _ Std_JSON_ZeroCopy_Serializer.Dummy__
var _ Std_JSON_ZeroCopy_Deserializer_Core.Dummy__
var _ Std_JSON_ZeroCopy_Deserializer_Strings.Dummy__
var _ Std_JSON_ZeroCopy_Deserializer_Numbers.Dummy__

type Dummy__ struct{}


// Definition of class Default__
type Default__ struct {
  dummy byte
}

func New_Default___() *Default__ {
  _this := Default__{}

  return &_this
}

type CompanionStruct_Default___ struct {
}
var Companion_Default___ = CompanionStruct_Default___ {
}

func (_this *Default__) Equals(other *Default__) bool {
  return _this == other
}

func (_this *Default__) EqualsGeneric(x interface{}) bool {
  other, ok := x.(*Default__)
  return ok && _this.Equals(other)
}

func (*Default__) String() string {
  return "Std_JSON_ZeroCopy_Deserializer_ObjectParams.Default__"
}
func (_this *Default__) ParentTraits_() []*_dafny.TraitID {
  return [](*_dafny.TraitID){};
}
var _ _dafny.TraitOffspring = &Default__{}

func (_static *CompanionStruct_Default___) Colon(cs Std_JSON_Utils_Cursors.Cursor__) Std_Wrappers.Result {
  var _664_valueOrError0 Std_Wrappers.Result = (cs).AssertChar(_dafny.CodePoint(':'))
  _ = _664_valueOrError0
  if ((_664_valueOrError0).IsFailure()) {
    return (_664_valueOrError0).PropagateFailure()
  } else {
    var _665_cs Std_JSON_Utils_Cursors.Cursor__ = (_664_valueOrError0).Extract().(Std_JSON_Utils_Cursors.Cursor__)
    _ = _665_cs
    return Std_Wrappers.Companion_Result_.Create_Success_((_665_cs).Split())
  }
}
func (_static *CompanionStruct_Default___) KeyValueFromParts(k Std_JSON_Utils_Cursors.Split, colon Std_JSON_Utils_Cursors.Split, v Std_JSON_Utils_Cursors.Split) Std_JSON_Utils_Cursors.Split {
  var _666_sp Std_JSON_Utils_Cursors.Split = Std_JSON_Utils_Cursors.Companion_Split_.Create_SP_(Std_JSON_Grammar.Companion_JKeyValue_.Create_KeyValue_((k).Dtor_t().(Std_JSON_Grammar.Jstring), (colon).Dtor_t().(Std_JSON_Grammar.Structural), (v).Dtor_t().(Std_JSON_Grammar.Value)), (v).Dtor_cs())
  _ = _666_sp
  return _666_sp
}
func (_static *CompanionStruct_Default___) ElementSpec(t Std_JSON_Grammar.JKeyValue) _dafny.Sequence {
  return Std_JSON_ConcreteSyntax_Spec.Companion_Default___.KeyValue(t)
}
func (_static *CompanionStruct_Default___) Element(cs Std_JSON_Utils_Cursors.Cursor__, json func (Std_JSON_Utils_Cursors.Cursor__) Std_Wrappers.Result) Std_Wrappers.Result {
  var _667_valueOrError0 Std_Wrappers.Result = Std_JSON_ZeroCopy_Deserializer_Strings.Companion_Default___.String(cs)
  _ = _667_valueOrError0
  if ((_667_valueOrError0).IsFailure()) {
    return (_667_valueOrError0).PropagateFailure()
  } else {
    var _668_k Std_JSON_Utils_Cursors.Split = (_667_valueOrError0).Extract().(Std_JSON_Utils_Cursors.Split)
    _ = _668_k
    var _669_p func (Std_JSON_Utils_Cursors.Cursor__) Std_Wrappers.Result = Companion_Default___.Colon
    _ = _669_p
    var _670_valueOrError1 Std_Wrappers.Result = Std_JSON_ZeroCopy_Deserializer_Core.Companion_Default___.Structural((_668_k).Dtor_cs(), func (coer44 func (Std_JSON_Utils_Cursors.Cursor__) Std_Wrappers.Result) func (Std_JSON_Utils_Cursors.Cursor__) Std_Wrappers.Result {
      return func (arg46 Std_JSON_Utils_Cursors.Cursor__) Std_Wrappers.Result {
        return coer44(arg46)
      }
    }(_669_p))
    _ = _670_valueOrError1
    if ((_670_valueOrError1).IsFailure()) {
      return (_670_valueOrError1).PropagateFailure()
    } else {
      var _671_colon Std_JSON_Utils_Cursors.Split = (_670_valueOrError1).Extract().(Std_JSON_Utils_Cursors.Split)
      _ = _671_colon
      var _672_valueOrError2 Std_Wrappers.Result = ((json))((_671_colon).Dtor_cs())
      _ = _672_valueOrError2
      if ((_672_valueOrError2).IsFailure()) {
        return (_672_valueOrError2).PropagateFailure()
      } else {
        var _673_v Std_JSON_Utils_Cursors.Split = (_672_valueOrError2).Extract().(Std_JSON_Utils_Cursors.Split)
        _ = _673_v
        var _674_kv Std_JSON_Utils_Cursors.Split = Companion_Default___.KeyValueFromParts(_668_k, _671_colon, _673_v)
        _ = _674_kv
        return Std_Wrappers.Companion_Result_.Create_Success_(_674_kv)
      }
    }
  }
}
func (_static *CompanionStruct_Default___) OPEN() uint8 {
  return uint8(_dafny.CodePoint('{'))
}
func (_static *CompanionStruct_Default___) CLOSE() uint8 {
  return uint8(_dafny.CodePoint('}'))
}
// End of class Default__
